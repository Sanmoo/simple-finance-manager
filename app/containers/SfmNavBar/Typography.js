import styled from 'styled-components';
import { Typography as TB } from '@material-ui/core';

const Typography = styled(TB)`
  flex-grow: 1;
`;

export default Typography;
