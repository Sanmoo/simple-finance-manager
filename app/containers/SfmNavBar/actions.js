import {
  SET_DRAWER_OPEN,
  COPY_SPREADSHEET_LINK,
  GO_TO_SPREADSHEET,
} from './constants';

export function setDrawerOpen(open) {
  return {
    type: SET_DRAWER_OPEN,
    open,
  };
}

export function copySpreadsheetLink(link) {
  return {
    type: COPY_SPREADSHEET_LINK,
    link,
  };
}

export function goToSpreadsheet(link) {
  return {
    type: GO_TO_SPREADSHEET,
    link,
  };
}
