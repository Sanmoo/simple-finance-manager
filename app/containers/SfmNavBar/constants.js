/*
 *
 * SfmNavBar constants
 *
 */

export const DEFAULT_ACTION = 'app/SfmNavBar/DEFAULT_ACTION';
export const SET_DRAWER_OPEN = 'app/SfmNavBar/SET_DRAWER_OPEN';
