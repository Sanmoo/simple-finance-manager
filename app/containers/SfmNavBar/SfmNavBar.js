/**
 *
 * SfmNavBar
 *
 */

import React, { useCallback, useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { AppBar, Toolbar, IconButton } from '@material-ui/core';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import { withStyles } from '@material-ui/core/styles';
import AccountCircle from '@material-ui/icons/AccountCircle';
import ConfirmationDialog from 'components/ConfirmationDialog/Loadable';
import ChevronLeftIcon from '@material-ui/icons/ChevronLeft';
import Divider from '@material-ui/core/Divider';
import ChevronRightIcon from '@material-ui/icons/ChevronRight';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import Drawer from '@material-ui/core/Drawer';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import ReceiptIcon from '@material-ui/icons/Receipt';
import DashboardIcon from '@material-ui/icons/Dashboard';
import MenuIcon from '@material-ui/icons/Menu';
import Typography from './Typography';

const drawerWidth = 240;

const styles = theme => ({
  drawer: {
    width: drawerWidth,
    flexShrink: 0,
  },
  drawerPaper: {
    width: drawerWidth,
  },
  drawerHeader: {
    display: 'flex',
    alignItems: 'center',
    padding: '0 8px',
    ...theme.mixins.toolbar,
    justifyContent: 'flex-end',
  },
  content: {
    flexGrow: 1,
    padding: theme.spacing.unit * 3,
    transition: theme.transitions.create('margin', {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen,
    }),
    marginLeft: -drawerWidth,
  },
  contentShift: {
    transition: theme.transitions.create('margin', {
      easing: theme.transitions.easing.easeOut,
      duration: theme.transitions.duration.enteringScreen,
    }),
    marginLeft: 0,
  },
});

const SfmNavBar = ({
  onSignOff,
  title,
  classes,
  drawerOpen,
  onDrawerOpen,
  onDrawerClose,
  theme,
  onDashboardClick,
  onEntriesClick,
}) => {
  const [anchorEl, setAnchorEl] = useState(null);

  const [signOffDialogOpened, setSignOffDialogOpened] = useState(false);
  const onSignOffMenuClicked = useCallback(
    () => setSignOffDialogOpened(true),
    [],
  );
  const onSignOffDialogDismiss = useCallback(
    () => setSignOffDialogOpened(false),
    [],
  );
  const onSignOffDialogConfirm = useCallback(() => {
    setSignOffDialogOpened(false);
    onSignOff();
  }, [onSignOff]);

  const userMenuOpened = Boolean(anchorEl);

  if (!title) {
    return <div />;
  }

  return (
    <React.Fragment>
      <AppBar position="static">
        <Toolbar>
          <IconButton
            color="inherit"
            aria-label="Open drawer"
            onClick={onDrawerOpen}
            className={classNames(
              classes.menuButton,
              drawerOpen && classes.hide,
            )}
          >
            <MenuIcon />
          </IconButton>
          <Typography variant="h6" color="inherit">
            {title}
          </Typography>
          <div>
            <IconButton
              aria-owns={userMenuOpened ? 'menu-appbar' : undefined}
              aria-haspopup="true"
              onClick={event => setAnchorEl(event.currentTarget)}
              color="inherit"
            >
              <AccountCircle />
            </IconButton>
            <Menu
              id="menu-appbar"
              anchorEl={anchorEl}
              anchorOrigin={{
                vertical: 'top',
                horizontal: 'right',
              }}
              transformOrigin={{
                vertical: 'top',
                horizontal: 'right',
              }}
              open={userMenuOpened}
              onClose={() => setAnchorEl(null)}
            >
              <MenuItem
                onClick={() => {
                  setAnchorEl(null);
                  onSignOffMenuClicked();
                }}
              >
                Sign off
              </MenuItem>
            </Menu>
          </div>
        </Toolbar>
      </AppBar>
      <Drawer
        className={classes.drawer}
        variant="persistent"
        anchor="left"
        open={drawerOpen}
        classes={{
          paper: classes.drawerPaper,
        }}
      >
        <div className={classes.drawerHeader}>
          <IconButton onClick={onDrawerClose}>
            {theme.direction === 'ltr' ? (
              <ChevronLeftIcon />
            ) : (
              <ChevronRightIcon />
            )}
          </IconButton>
        </div>
        <Divider />
        <List>
          {['Dashboard', 'Entradas'].map((text, index) => (
            <ListItem
              button
              key={text}
              onClick={() => {
                if (index === 0) {
                  onDashboardClick();
                } else if (index === 1) {
                  onEntriesClick();
                } else {
                  throw new Error('Assert false');
                }
                onDrawerClose();
              }}
            >
              <ListItemIcon>
                {index === 0 ? <DashboardIcon /> : <ReceiptIcon />}
              </ListItemIcon>
              <ListItemText primary={text} />
            </ListItem>
          ))}
        </List>
      </Drawer>
      <ConfirmationDialog
        opened={signOffDialogOpened}
        onDismiss={onSignOffDialogDismiss}
        onConfirm={onSignOffDialogConfirm}
        content="Tem certeza que deseja sair?"
      />
    </React.Fragment>
  );
};

SfmNavBar.propTypes = {
  onSignOff: PropTypes.func.isRequired,
  title: PropTypes.string,
  classes: PropTypes.object.isRequired,
  theme: PropTypes.object.isRequired,
  drawerOpen: PropTypes.bool.isRequired,
  onDrawerOpen: PropTypes.func.isRequired,
  onDrawerClose: PropTypes.func.isRequired,
  onDashboardClick: PropTypes.func.isRequired,
  onEntriesClick: PropTypes.func.isRequired,
};

export default withStyles(styles, { withTheme: true })(SfmNavBar);
