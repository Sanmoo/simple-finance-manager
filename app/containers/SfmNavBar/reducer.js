/*
 *
 * SfmNavBar reducer
 *
 */
import produce from 'immer';
import { DEFAULT_ACTION, SET_DRAWER_OPEN } from './constants';

export const initialState = {
  drawerOpen: false,
};

/* eslint-disable default-case, no-param-reassign */
const sfmNavBarReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case DEFAULT_ACTION:
        break;
      case SET_DRAWER_OPEN:
        draft.drawerOpen = action.open;
        break;
    }
  });

export default sfmNavBarReducer;
