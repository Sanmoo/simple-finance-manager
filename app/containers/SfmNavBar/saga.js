/**
 * Gets the repositories of the user from Github
 */

import { takeLatest, put } from 'redux-saga/effects';
import { popSnackbar } from 'containers/SnackbarContainer/actions';
import copy from 'copy-to-clipboard';
import { COPY_SPREADSHEET_LINK, GO_TO_SPREADSHEET } from './constants';

export function* handleCopySpreadsheetLink({ link }) {
  copy(link);
  yield put(popSnackbar(`link successfully copied to clipboard`));
}

export function* handleGoToSpreadsheet({ link }) {
  window.open(link, '_blank');
}

export default function* saga() {
  yield takeLatest(COPY_SPREADSHEET_LINK, handleCopySpreadsheetLink);
  yield takeLatest(GO_TO_SPREADSHEET, handleGoToSpreadsheet);
}
