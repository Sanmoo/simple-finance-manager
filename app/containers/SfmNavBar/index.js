/**
 *
 * SfmNavBar
 *
 */

import { connect } from 'react-redux';
import { compose } from 'redux';
import { signOff } from 'containers/App/actions';
import SfmNavBar from './SfmNavBar';

const withConnect = connect(
  null,
  { onSignOff: signOff },
);

export default compose(withConnect)(SfmNavBar);
