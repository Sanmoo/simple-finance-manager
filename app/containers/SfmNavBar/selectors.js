import { createSelector } from 'reselect';
import { makeSelectCurrentPage } from 'containers/App/selectors';
import { PAGES_TITLE_BAR } from 'utils/businessConstants';
import { initialState } from './reducer';

const selectsfmNavBarDomain = state => state.sfmNavBar || initialState;

const makeSelectBarTitle = () =>
  createSelector(
    makeSelectCurrentPage(),
    page => PAGES_TITLE_BAR[page],
  );

const makeSelectDrawerOpen = () =>
  createSelector(
    selectsfmNavBarDomain,
    substate => substate.drawerOpen,
  );

export { makeSelectBarTitle, makeSelectDrawerOpen };
