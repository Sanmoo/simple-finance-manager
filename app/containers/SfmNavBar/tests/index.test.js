/**
 *
 * Tests for SfmNavBar
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render } from 'react-testing-library';
// import 'jest-dom/extend-expect'; // add some helpful assertions

import SfmNavBar from '../SfmNavBar';

describe('<SfmNavBar />', () => {
  let props;

  beforeEach(() => {
    props = {
      onSignOff: jest.fn(),
      title: 'This is my title',
      drawerOpen: false,
      onDrawerOpen: jest.fn(),
      onDrawerClose: jest.fn(),
      onDashboardClick: jest.fn(),
      onEntriesClick: jest.fn(),
      onGoToSpreadsheetClick: jest.fn(),
      onCopySpreadsheetLinkClicked: jest.fn(),
      spreadsheetUrl: 'url-test',
    };
  });

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(<SfmNavBar {...props} />);
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(<SfmNavBar {...props} />);
    expect(firstChild).toMatchSnapshot();
  });
});
