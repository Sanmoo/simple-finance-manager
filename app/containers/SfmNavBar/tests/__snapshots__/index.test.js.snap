// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SfmNavBar /> Should render and match the snapshot 1`] = `
<header
  class="MuiPaper-root-179 MuiPaper-elevation4-185 MuiAppBar-root-170 MuiAppBar-positionStatic-174 MuiAppBar-colorPrimary-177"
>
  <div
    class="MuiToolbar-root-206 MuiToolbar-regular-208 MuiToolbar-gutters-207"
  >
    <button
      aria-label="Open drawer"
      class="MuiButtonBase-root-216 MuiIconButton-root-210 MuiIconButton-colorInherit-211"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label-215"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root-219"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root-322"
      />
    </button>
    <h6
      class="MuiTypography-root-228 MuiTypography-h6-245 MuiTypography-colorInherit-257 Typography-sc-1tw305x-0 icPXXg"
    >
      This is my title
    </h6>
    <div>
      <button
        aria-haspopup="true"
        class="MuiButtonBase-root-216 MuiIconButton-root-210 MuiIconButton-colorInherit-211"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label-215"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-219"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root-322"
        />
      </button>
    </div>
  </div>
</header>
`;
